torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -0,0 +1,1093 @@
+#include <torch/csrc/jit/tensorexpr/ir_simplifier.h>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+
+SimplifierHashType Term::hashVars() const {
+  SimplifierHashType hash = 0;
+  for (auto* v : variables_) {
+    hash = hasher_.hash_combine(hash, hasher_.hash(v));
+  }
+
+  return hash;
+}
+
+void Term::sort() {
+  // order of ops important for float
+  if (dtype().is_floating_point()) {
+    throw std::logic_error("reordering FP ops");
+  }
+  std::sort(
+      variables_.begin(), variables_.end(), [&](const Expr* a, const Expr* b) {
+        return hasher_.hash(a) < hasher_.hash(b);
+      });
+}
+
+SimplifierHashType Polynomial::hashVars() const {
+  SimplifierHashType hash = 0;
+  for (auto* v : variables_) {
+    hash = hasher_.hash_combine(hash, hasher_.hash(v));
+  }
+  return hash;
+}
+
+void Polynomial::sort() {
+  if (dtype().is_floating_point()) {
+    throw std::logic_error("reordering FP ops");
+  }
+  std::sort(
+      variables_.begin(), variables_.end(), [&](const Expr* a, const Expr* b) {
+        return hasher_.hash(a) < hasher_.hash(b);
```

Comment:
It is generally a bad idea to sort on hash, as we've already discovered - it can make tests brittle and can in future lead to non-deterministic behavior (even if it's not like that now). I suggest following up on that - a better way would be to normalize/canonicalize the expression - i.e. define a strict order how sub-expressions should be placed in the parent expression.

Reply:
The hash defines an ordering though, doesn't it. The issues we've been having in this diff is that the hash wasn't consistent across platforms.

I'm a bit resistant to some kind of adhoc ordering because in the case of compound opaque objects I think it can get complicated. Am open to suggestions on this though.

Reply:
You're not wrong though, I found a bug in the hash and it changes the order of a bunch of things.